lib/datadog/search-syntax/src/builder.rs

```diff
@@ -0,0 +1,514 @@
+                                make_node(query),
+                                if *lower_inclusive {
+                                    ast::Opcode::Ge
+                                } else {
+                                    ast::Opcode::Gt
+                                },
+                                make_node(ast::Expr::Literal(make_node(lower.clone().into()))),
+                            ))
+                        }
+                        // Definitive range.
+                        _ => {
+                            self.coalesce();
+                            make_container_group(make_op(
+                                make_node(make_op(
+                                    make_node(query.clone()),
+                                    if *lower_inclusive {
+                                        ast::Opcode::Ge
+                                    } else {
+                                        ast::Opcode::Gt
+                                    },
+                                    make_node(ast::Expr::Literal(make_node(lower.clone().into()))),
+                                )),
+                                ast::Opcode::And,
+                                make_node(make_op(
+                                    make_node(query),
+                                    if *upper_inclusive {
+                                        ast::Opcode::Le
+                                    } else {
+                                        ast::Opcode::Lt
+                                    },
+                                    make_node(ast::Expr::Literal(make_node(upper.clone().into()))),
+                                )),
+                            ))
+                        }
+                    }
+                })
+                .collect(),
+            // Negation. If the node is an operation type, wrap in a container before negating.
+            QueryNode::NegatedNode { node } => {
+                let expr = recurse(self.parse_node(node).into_iter());
+
+                let node = match expr {
+                    ast::Expr::Op(_) => make_container_group(expr),
+                    _ => expr,
+                };
+
+                vec![make_not(node)]
+            }
+            // Compound.
+            QueryNode::Boolean { oper, nodes } => {
+                let exprs = nodes
+                    .iter()
+                    .map(|node| recurse(self.parse_node(node).into_iter()));
+
+                vec![recurse_op(exprs, oper)]
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::{compile, parse};
+    use vrl_parser::ast;
+
+    // Lhs = Datadog syntax. Rhs = VRL equivalent.
+    static TESTS: &[(&str, &str)] = &[
+        // Match everything (empty).
+        ("", "(exists(.message) || (exists(.custom.error.message) || (exists(.custom.error.stack) || (exists(.custom.title) || exists(._default_)))))"),
```

Comment:
Here are some found discrepancies with logs-backend(confirmed with unit tests in logs-backend):

"", "*:*"  - should match really anything without checking any fields.
"_exists_:a" - matches tags ["a:b"], but doesn't match tags ["a"](weird though, agree) - this is apparently documented here https://docs.datadoghq.com/logs/search_syntax/#tags (so if tags don't follow tag:value syntax we can search them only by "tags:a"). Similarly with `_missing_:a`.
a:*bla - I think here we should check like ends_with(tags.a, "bla") or using regex without word boundaries, otherwise something like tags: ["a:abc-bla-abc"] would pass. Tag values and should not be analyzed as text values(only `message`, `custome.error.stack` ..). Similarly with prefixes like `b:bla*` and custom fields `@a:*bla`.
[1 TO 10] - default fields are analysed as text values therefore it will do comparison in the alphabetical order. For example, ‘message:3’ will be matched by ‘message:[1 TO 9]’, but won’t be matched by `message:[1 TO 10]`.  Similarly for wildcard queries like `[* TO 50]`
a:[1 TO 10] - similarly here - tags are also analysed as string values.
@b:[1 TO 10] - will match only numerical values, for example, will match `custom.b:5`, but won’t match `custom.b:“5”`

Reply:
@vladimir-dd, to confirm one small point... you have `:` listed as an example. Did you mean `*:*`? or should literally `:` also work?

Right now, `:` appears to be a parse error in the grammar.

Reply:
ah, sorry, I meant `*:*`. Throwing a grammar error in case of `:` is correct.